File: php/binary/binary.php
<?php

function parse_binary(string $binary) : int {
   $decimal = (int) 0;
   $exp = strlen($binary) - 1;
   
   foreach (str_split($binary) as $bit) {
      if ($bit != '1' && $bit != '0') {
         throw new InvalidArgumentException(
            'Expected binary bit. Argument contained: '.$bit);
      }
      $decimal += $bit << $exp;
      --$exp;
   }
   
   return $decimal;
}

File: c/binary/src/binary.c
#include "binary.h" /* #def INVALID */

int convert(char const* binary) {
  int decimal = 0;
  for (char c = binary[0]; c != '\0'; c = *(++binary)) {
    decimal <<= 1;
    if      (c == '1') decimal |= 1;
    else if (c != '0') return INVALID;
  }
  return decimal;
}
